File: src/main/java/frc/team687/utilities/statespace/StateSpaceController.java
package frc.team687.utilities.statespace;

import Jama.Matrix;

public class StateSpaceController {

    private StateSpaceGains[] m_gains;
    private Matrix m_U_min, m_U_max;
    private StateSpaceObserver m_observer;

    private int m_selectedGainsIndex;

    public StateSpaceController(StateSpaceGains[] gains, Matrix U_min, Matrix U_max) {
        this.m_gains = gains;
        this.m_U_min = U_min;
        this.m_U_max = U_max;
        this.m_selectedGainsIndex = 0;
    }

    public void setGainsIndex(int index) {
        assert index < this.m_gains.length : "Gains index must be in bounds";
        this.m_selectedGainsIndex = index;
    }

    public Matrix getFeedforwardOutput(Matrix nextGoal, Matrix estimatedState) {
        StateSpaceGains currentGains = this.m_gains[this.m_selectedGainsIndex];

        // Discrete-time reference tracking feedforward input: u_ff[k] = pinv(B) * (x[k+1] - Ax^[k]),
        // where x[k+1] = reference
        Matrix Uff = currentGains.Kff.times(nextGoal.plus(currentGains.A.times(estimatedState)));

        // Discrete-time, reference tracking, not including feedforward input: u_c = -K * (x^ - r) = K * (r - x^)
        Matrix Uc = currentGains.K.times(nextGoal.minus(estimatedState));

        // u = uc + uff
        return Uc.plus(Uff);
    }

    public Matrix getReferenceTrackingOutput(Matrix reference, Matrix estimatedState) {
        StateSpaceGains currentGains = this.m_gains[this.m_selectedGainsIndex];

        Matrix referenceInput = currentGains.N.times(reference);


        Matrix controlLawInput = currentGains.K.times(estimatedState);

        // u = -K*x + N*r
        referenceInput.minusEquals(controlLawInput);
        JamaUtils.printMatrix(estimatedState);

        return controlLawInput;
    }

    public Matrix boundOutput(Matrix output) {
        return JamaUtils.boundMatrix(output, this.m_U_min, this.m_U_max);
    }

    public Matrix getBoundedOutput(Matrix reference, Matrix estimatedState) {
        return this.boundOutput(getReferenceTrackingOutput(reference, estimatedState));
    }

}

File: src/main/python/utilities/state_space/state_space_controller.py
import numpy as np
from utilities.state_space.state_space_gains import GainsList


class StateSpaceController(object):

    def __init__(self, gains: GainsList, u_initial, r_initial, u_max, u_min):
        self.gains = gains
        self.gains_index = 0
        self.current_gains = self.gains.get_gains(self.gains_index)

        self.u_max = u_max
        self.u_min = u_min

        self.u = u_initial
        self.r = r_initial

    def set_index(self, index: int):
        self.gains_index = index
        self.current_gains = self.gains.get_gains(index)

    def update_ff(self, mot_prof, estimated_state):
        raise NotImplementedError

    def bounded_update_ff(self, r, x_hat):
        return np.clip(self.update_ff(r, x_hat), self.u_min, self.u_max)

    def update_ff(self, r, x_hat):
        gains = self.current_gains

        uff = gains.Kff * (r - (gains.A * self.r))
        self.u = gains.K * (r - x_hat)

        return self.u + uff

    def bounded_update(self, r, x_hat):
        return np.clip(self.update(r, x_hat), self.u_min, self.u_max)

    def update(self, r, x_hat):
        gains = self.current_gains

        self.r = r
        self.u = gains.K * (r - x_hat)

        return self.u


